src/particle_filter.cpp

```diff
@@ -1,33 +1,95 @@
 #include "particle_filter.h"
 
 #include <algorithm>
+#include <cmath>
+
+#include <ros/console.h>
 
 // ----------------------------------------------------------------------------------------------------
 
-ParticleFilter::ParticleFilter() : i_current_(0)
+ParticleFilter::ParticleFilter() : min_samples_(0), max_samples_(0), kld_err_(0), kld_z_(0), alpha_slow_(0),
+    alpha_fast_(0), w_slow_(0), w_fast_(0), i_current_(0), kd_tree_(nullptr)
 {
 }
 
 // ----------------------------------------------------------------------------------------------------
 
+void ParticleFilter::configure(tue::Configuration config)
+{
+    // Extra variable needed as tue::Configuration doesn't support
+    // unisgned interters
+    int min, max;
+    config.value("min_particles", min);
+    config.value("max_particles", max);
+    min_samples_ = min;
+    max_samples_ = max;
+
+    kld_err_ = 0.01;
+    kld_z_ = 0.99;
+    config.value("kld_err", kld_err_, tue::config::OPTIONAL);
+    config.value("kld_z", kld_z_, tue::config::OPTIONAL);
+
+    alpha_slow_ = 0;
+    alpha_fast_ = 0;
+    config.value("recovery_alpha_slow", alpha_slow_, tue::config::OPTIONAL);
+    config.value("recovery_alpha_fast", alpha_fast_, tue::config::OPTIONAL);
+
+    std::array<double, 3> cell_size = {0.5, 0.5, 10*M_PI/180};
+    config.value("cell_size_x", cell_size[0], tue::config::OPTIONAL);
+    config.value("cell_size_y", cell_size[1], tue::config::OPTIONAL);
+    config.value("cell_size_theta", cell_size[2], tue::config::OPTIONAL);
+
+    samples_[0].reserve(max_samples_);
+    samples_[1].reserve(max_samples_);
+
+    cluster_cache_.reserve(max_samples_);
+
+    limit_cache_.clear();
+    limit_cache_.resize(max_samples_, 0);
+
+    kd_tree_.reset(new KDTree(max_samples_, cell_size));
+
+    ROS_INFO_STREAM("[ED Localization] min_samples: " << min_samples_ << ", max_samples: " << max_samples_);
```

Comment:
Maybe put it in one call? Thinking about it, I actually believe that adding '[ED Localization]' is bad practice and is often done because we hardly ever use proper debugging tools...